### PR TITLE
Final touches on filter Classes

### LINKED ISSUE #36: Filter state not properly supported in filter class objects
## Description

The current Filter implementation does not support setting the state of the filter properly.
Filtering without saving the state is not affected.
Related to #33

### PR DESCRIPTION
**Pull request by mberz**
### Which issue(s) are closed by this pull request?

Closes #36 , #102 and #204   

### Changes proposed in this pull request:

- Implement the saving of filter states for filter objects
- rename `filter.shape` to `filter.n_channels`
- add `__repr__` strings
- improve documentation and add separate concepts for Filters
- make helper functions private

### CODE REVIEW

**Comment by sikersten** [comment]
Regarding the new introduction of concepts in the docs, I'd suggest to change the structure a little. Instead of having the files pyfar.classes_...rst and pyfar.concepts_....rst we can add subfolders _classes_ and _concepts_ to the docs folder. The files can be imported into the corresponding files _classes.rst_ and _concepts.rst_, e.g, via _classes/filter.rst_.

File: pyfar/classes/filter_concepts.py [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+
+    in = pf.Signal([1, 2, 3, 4], 44100)
+    out = filter.process(in)
+
+The output is ``out.time = [[2, 2, 2, 2]]`` and has the same number of samples
+as the input.
+
+The output will be the same no matter how often ``process`` is called. This
+default behavior is often desired. In some cases, a different functionality can
+be usefull. For blockwise processing of input signals, the Filter object can
```
Comment (sikersten) [inline pyfar/classes/filter_concepts.py:45]:
```suggestion
be useful. For blockwise processing of input signals, the Filter object can
```

File: tests/test_filter_classes.py [hunk 2]
```diff
@@ -183,21 +334,42 @@ def test_filter_sos_process_multi_dim_filt(impulse):
...
+    # filter in two blocks with correct handling of the state
+    Filter.init_state(signal.cshape, 'zeros')
+    block_a = Filter.process(signal[0, :3], reset=False)
+    block_b = Filter.process(signal[0, 3:], reset=False)
+    # outputs have to be identical in this case
+    npt.assert_array_equal(complete[0, :3].time, block_a.time)
+    npt.assert_array_equal(complete[0, 3:].time, block_b.time)
+
+
 def test_atleast_3d_first_dim():
```
Comment (sikersten) [inline tests/test_filter_classes.py:358]: Not really necessary, but could be changed:
```suggestion
def test__atleast_3d_first_dim():
```
Could be applied to functions below accordingly

**Review by sikersten — COMMENTED** [review]
Only minor comments, all in all it looks good to me :)

**Comment by sikersten** [comment]
I realized it gets a little messy when introducing the concepts in this pull request due to the large number of file changes. Therefore, I opened a different pull request #243 for that, so we can focus on the improvement w.r.t. the filter functionality in this pull.

File: pyfar/classes/filter.py [hunk 3]
```diff
@@ -89,24 +131,29 @@ def __init__(
...
-    def initialize(self):
-        raise NotImplementedError("Abstract class method")
+    def init_state(self, state='zeros'):
+        self._state = state
+        self._initialized = True
+
+    @property
+    def coefficients(self):
+        """Coefficients of the filter"""
+        self._coefficients
```
Comment (al-de) [inline pyfar/classes/filter.py:156]:
```suggestion
        return self._coefficients
```
I think the `_coefficients` need to be returned for the property.

File: pyfar/classes/filter_concepts.py [hunk 4]
```diff
@@ -0,0 +1,87 @@
+"""
+Filter Types
+------------
+
+There are three types of Filter objects:
+
+- **FIR Filter:** Finite Impulse Response (FIR) filters are non-recursive
+  filters. FIR filters very flexible and can have arbitrary magnitude and
```
Comment (al-de) [inline pyfar/classes/filter_concepts.py:8]:
```suggestion
  filters. FIR filters are very flexible and can have arbitrary magnitude and
```

File: pyfar/classes/filter_concepts.py [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+  phase responses.
+- **IIR Filter:** Infinite Impulse Response (IIR) filters are recursive
+  filters. The can achieve steeper filter slopes than FIR filters of the same
```
Comment (al-de) [inline pyfar/classes/filter_concepts.py:11]:
```suggestion
  filters. They can achieve steeper filter slopes than FIR filters of the same
```

File: pyfar/classes/filter_concepts.py [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+
+The above initializes the state with zeros, and if the filter is called with
+blocks of the input
+
+.. code-block:: python
+
+    block_one = filter.process(in[0, 0:2])
+    block_two = filter.process(in[0, 2:4])
+
+the the blockwise output yields the same as the complete output ``out`` seen
```
Comment (al-de) [inline pyfar/classes/filter_concepts.py:61]:
```suggestion
the blockwise output yields the same as the complete output ``out`` seen
```

**Review by al-de — APPROVED** [review]
Thanks a lot for this huge implementation, only had a few minor comments. :)
Besides that everything looks fine to me.

File: pyfar/classes/filter_concepts.py (hunk 6 above)
Comment (f-brinkmann) [reply pyfar/classes/filter_concepts.py]: thanks, changed

File: pyfar/classes/filter_concepts.py (hunk 5 above)
Comment (f-brinkmann) [reply pyfar/classes/filter_concepts.py]: thanks, changed

File: pyfar/classes/filter_concepts.py (hunk 4 above)
Comment (f-brinkmann) [reply pyfar/classes/filter_concepts.py]: thanks, changed

File: pyfar/classes/filter.py (hunk 3 above)
Comment (f-brinkmann) [reply pyfar/classes/filter.py]: Yes, changed!
**NOTE THE FOLLOWING:** `FilterFIR` also stores a-coefficients internally, that should not be visible to the user. I thus overwrote the property for `FilterFIR` and added additional tests for that.

**Review by sikersten — APPROVED** [review]

**Review by al-de — APPROVED** [review]

**Pull request merged**